File: DKPlugins/DKDuktape/DKFocusEvent.cpp
/*
* This source file is part of digitalknob, the cross-platform C/C++/Javascript/Html/Css Solution
*
* For the latest information, see https://github.com/aquawicket/DigitalKnob
*
* Copyright(c) 2010 - 2023 Digitalknob Team, and contributors
*
* Permission is hereby granted, free of charge, to any person obtaining a copy
* of this software and associated documentation files(the "Software"), to deal
* in the Software without restriction, including without limitation the rights
* to use, copy, modify, merge, publish, distribute, sublicense, and /or sell
* copies of the Software, and to permit persons to whom the Software is
* furnished to do so, subject to the following conditions :
*
* The above copyright notice and this permission notice shall be included in all
* copies or substantial portions of the Software.
*
* THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
* IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
* FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
* AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
* LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
* OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
* SOFTWARE.
*/

// https://developer.mozilla.org/en-US/docs/Web/API/FocusEvent

#include "DK/DKApp.h"
#include "DKDuktape/DKEvent.h"
#include "DKDuktape/DKEventTarget.h"
#include "DKDuktape/DKFocusEvent.h"
#include "EventTest/ConsoleInput.h"


bool DKFocusEvent::Init(){
	DKDEBUGFUNC();

	 //// Instance properties ////
	DKDuktape::AttachFunction("CPP_DKFocusEvent_relatedTarget", DKFocusEvent::relatedTarget); //Read only

	DKClass::DKCreate("DKDuktape/DKFocusEvent.js");
	return true;
}


 //// Instance properties ////
int DKFocusEvent::relatedTarget(duk_context* ctx){ //Read only
	DKString eventAddress = duk_require_string(ctx, 0);
	ConsoleInput* event = (ConsoleInput*)DKDuktape::addressToPointer(eventAddress);
	if (!event) {
		DKERROR("event invalid! \n");
		duk_push_undefined(ctx);
		return true;
	}
	duk_push_int(ctx, event->relatedTarget);
	return true;
}

File: DKPlugins/DKDuktape/DKFocusEvent.js
// [MDN] https://developer.mozilla.org/en-US/docs/Web/API/FocusEvent
// [INTERFACE] https://w3c.github.io/uievents/#interface-focusevent

// [FocusEvent()] https://developer.mozilla.org/en-US/docs/Web/API/FocusEvent/FocusEvent
var FocusEvent = function(_type, _options, _pointer) {
	
    ////// Instance properties //////
	// [FocusEvent.relatedTarget](Read only) https://developer.mozilla.org/en-US/docs/Web/API/FocusEvent/relatedTarget
    Object.defineProperty(this, "relatedTarget", {
        get: function relatedTarget() {
            return CPP_DKFocusEvent_relatedTarget(_pointer)
        }
    })
    
	////// toString //////
	if(this.toString() === "[object Object]"){
		this.toString = function(){
			return "[object FocusEvent]"
		}
	}
	
    return UIEvent.call(this, _type, _options, _pointer)
}
FocusEvent.prototype = UIEvent.prototype


// Called from C++ to dispatch Focus events
////////////////////////////////////////////////////////
function dispatchFocusEvent(_type, _options, _pointer) {
	const event = new FocusEvent(_type, _options, _pointer)
	event.target = new EventTarget(_pointer)
	event.target.dispatchEvent(event)
}
